File: replace-all-digits-with-characters-2.ts
// declare function replaceDigits, which takes in a string s, and returns a string.
function replaceDigits(s: string): string {
  // create mutable string variable shifted, and initialize it to an empty string.
  let shifted: string = "";
  // loop from 0 to the length of s.
  for (let i = 0; i < s.length; i++) {
    // if i is an even number.
    if (i % 2 === 0) {
      // set shifted to shifted plus the value of s at index i
      shifted += s[i];
    // else, if i is an odd number
    } else {
      // set shifted to shifted plus the character that gets return from a given charCode.
      // to get that charCode:
      // we get the charCode for the last character on shifted.
      // we also get the string integer from s at index i, and convert it to a number.
      // we add these two values together to get the charCode we need.
      shifted += String.fromCharCode(
        shifted.charCodeAt(shifted.length - 1) + parseInt(s[i])
      );
    }
  }
  return shifted;
}

File: replace-all-digits-with-characters-2.py
# define class Solution
class Solution:
    # define function replaceDigits, which takes in the class instance (self), string s, and
    # returns a string.
    def replaceDigits(self, s: str) -> str:
        # create variable alpha, which is a list of all alphabetical characters in order.
        alpha = list("abcdefghijklmnopqrstuvwxyz")
        # create variable res and initialize as an empty string
        res = ""
        # create variable i and initialize as 0
        i = 0

        # define function current, which takes in letter
        def current(letter):
            # define variable i and initialize as 0
            i = 0
            # while letter is not equal to the value of alpha at index i
            while letter != alpha[i]:
                # increment i by 1
                i += 1
            return i
        # while i is less than the length of s
        while i < len(s):
            # set res equal to res plus the value of s at index i
            res += s[i]
            # if i plus 1 is less than the length of s
            if (i + 1) < len(s):
                # set res equal to res plus the value of alpha at the return value of current with
                # the value of s at index i passed in, plus the value of s at index i plus 1,
                # converted to an integer.
                res += alpha[current(s[i]) + int(s[i+1])]

            # increment i by 2.
            i += 2
        return res
